src/bls/bls_aggregator.cpp

```diff
@@ -591,183 +663,262 @@ bls_rewards_response bls_aggregator::rewards_request(const address& addr, uint64
-            core.get_nettype(),
-            bls_aggregator::removal_type::liquidate,
-            removal_request.remove_pk,
-            removal_request.timestamp.count());
-    bls_signature sig = signer.signMsg(msg);
+    // NOTE: Validate the arguments
+    const bls_public_key& bls_pubkey = *maybe_bls_pubkey;
+    if (!bls_pubkey) {
+        throw oxen::traced<std::invalid_argument>(
+                "{} request for SN {} w/ the zero BLS pkey at height {} is invalid. Request "
+                "rejected"_format(label, pubkey, core.blockchain.get_current_blockchain_height()));
+    }
 
-    oxenc::bt_dict_producer d;
-    // BLS key of the node being liquidated:
-    d.append("liquidate", tools::view_guts(removal_request.remove_pk));
-    // signature of *this* snode of the liquidating pubkey:
-    d.append("signature", tools::view_guts(sig));
+    // NOTE: Serve the response from our cache if it's a repeated request
+    {
+        std::lock_guard lock{exit_liquidation_response_cache_mutex};
+        auto cache_it = exit_liquidation_response_cache.find(bls_pubkey);
+        if (cache_it != exit_liquidation_response_cache.end()) {
 
-    m.send_reply("200", std::move(d).str());
-}
+            const bls_exit_liquidation_cache_item& cache = cache_it->second;
+            const bls_exit_liquidation_response& response = type == bls_exit_type::normal ? cache.exit : cache.liquidation;
 
-// Common code for removal and liquidation requests, which only differ in three ways:
-// - the endpoint they go to;
-// - the tag that gets used in the msg_to_sign hash; and
-// - the key under which the signed pubkey gets confirmed back to us.
-bls_removal_liquidation_response bls_aggregator::removal_liquidation_request(const bls_public_key& bls_pubkey, removal_type type) {
-    // NOTE: The key in which the pubkey is inserted into the response which is a dictionary that is
-    // serialised onto OMQ with by all the service nodes we contacted.
-    std::string_view pubkey_key = "";
+            auto now_unix_ts = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch());
+            auto cache_ts = std::chrono::seconds(response.timestamp);
+
+            if (now_unix_ts >= cache_ts) {
+                std::chrono::seconds cache_age = now_unix_ts - cache_ts;
+                if (cache_age <= contract::REWARDS_SIGNATURE_EXPIRY) {
+                    log::trace(
+                            logcat,
+                            "Serving {} response from cache for SN {} (cached {})\n{}",
+                            label,
+                            pubkey,
+                            tools::get_human_readable_timespan(cache_age),
+                            dump_bls_exit_liquidation_response(response));
+                    return response;
+                }
+            }
+        }
+    }
 
     // NOTE: The OMQ endpoint to hit
+    bool removable = false;
```

Comment:
```suggestion
    bool removable;
```
to allow uninitialized value warnings if future code doesn't always set it in the following switch.